src/main/java/org/folio/rest/impl/PutOrderLineByIdHelper.java

```diff
@@ -161,14 +162,95 @@ public boolean isUpdateSuccessful() {
-      .thenAccept(itemIds -> {
-        // Temporal check. The idea is to create piece records for successfully created items and then throw exception
+      .thenCompose(itemIds -> {
         if (itemIds.size() != expectedItemsQuantity) {
           throw new IllegalStateException("Expected items quantity does not correspond to created items");
         }
-      });
+        return completedFuture(itemIds);
+      })
+      .thenCompose(itemIds -> createPieces(compPOL, expectedItemsQuantity, itemIds));
+  }
+
+  /**
+   * Creates Piece records corresponding to each item record associated with PO Line
+   *
+   * @param compPOL Composite PO line to update Inventory for
+   * @param expectedItemsQuantity Inventory items quantity
+   * @param itemIds List of item id
+   * @return CompletableFuture with newly created Pieces associated with PO line.
+   */
+  private CompletableFuture<Void> createPieces(PoLine compPOL, int expectedItemsQuantity, List<String> itemIds) {
+		CompletableFuture<Void> future = new VertxCompletableFuture<>(ctx);
+		if (itemIds.isEmpty()) {
+			future.complete(null);
+			return future;
+		}
+		List<CompletableFuture<Void>> futuresList = new ArrayList<>();
+		String poLineId = compPOL.getId();		
+		String endpoint = String.format(PIECES_ENDPOINT, poLineId);
+		
+		handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+		    .thenAccept(body -> {
+		    	int i=0;
+		      // No Pieces exists
+			    if (body.getInteger("total_records") == 0) { 
+				    while (i < expectedItemsQuantity) {
+					    futuresList.add(createPiece(poLineId, itemIds, i));
+					    i++;
+				    }
+			    } else if(body.getInteger("total_records") > 0) {
+			    	// if some of the piece records already exists out of expectedItemsQuantity then only create remaining pieces
+			    	int existingPieceRecordCount = body.getInteger("total_records");
+			    	int newQuantity = expectedItemsQuantity - existingPieceRecordCount;
+			    	while (i < newQuantity) {
+					    futuresList.add(createPiece(poLineId, itemIds, i));
+					    i++;
+				    }
+			    }
+			    logger.info("response from GET /pieces: " + body.encodePrettily());
+
+			    CompletableFuture.allOf(futuresList.toArray(new CompletableFuture[0]))
+			    .thenAccept(v -> future.complete(null))
+			    .exceptionally(t -> {
+				    future.completeExceptionally(t);
+				    return null;
+			    });
+		    });
+
+		return future;
   }
 
+  /**
+   * Construct Piece object associated with PO Line
+   *
+   * @param compPOL Composite PO line to update Inventory for
+   * @param itemIds List of itemIds that were created
+   * @param index index associated with each Piece to count total Pieces created
+   * @return CompletableFuture with new Piece record.
+   */
+  private CompletableFuture<Void> createPiece(String poLineId, List<String> itemIds, int index) {
+		// Construct Piece object
+		JsonObject pieceObj = new JsonObject();
```

Comment:
My personal preference is to use models generated based on json schemas. This allows to avoid mistakes and guaranty that any change in the schema results to compilation error rather then runtime exception. So the change is like following:
```java
    Piece pieceObj = new Piece();
		pieceObj.setPoLineId(poLineId);
		pieceObj.setItemId(itemId);
		pieceObj.setReceivingStatus(Piece.ReceivingStatus.EXPECTED);
...
			operateOnSubObj(HttpMethod.POST, resourcesPath(PIECES), JsonObject.mapFrom(pieceObj), httpClient, ctx, okapiHeaders, logger)
```

Reply:
Agree. Have made changes.